pennylane/transforms/batch_reduce.py

```diff
@@ -0,0 +1,302 @@
+                    ...
+                    return tapes, processing_fn
+
+            That is, the first argument must be the input QNode (or quantum tape) to transform, and
+            the function must return a tuple ``(list, function)`` containing:
+
+            * A list of new tapes to execute in a batch, and
+            * A processing function with signature ``processing_fn(List[float])``
+              which is applied to the flat list of results from the executed tapes.
+
+            If ``tapes`` is empty, then it is assumed no quantum evaluations
+            are required, and ``processing_fn`` will be passed an empty list.
+
+    Returns:
+        function: A hybrid quantum-classical function. Takes the same input arguments as
+        the input QNode, as well as two additional **experimental** keyword arguments:
+
+        * **batch_execute=False** (``bool``): If set to ``True``, the batch of tapes will
+          be executed via the QNode devices' ``batch_execute`` method. This can see significant
+          performance improvements on remote simulator and hardware devices. Note that
+          batch execution currently **does not support differentiability**.
+
+        * **parallel=False** (``bool``): If set to ``True``, the tapes are executed in parallel
+          using the `Dask <https://dask.org/>`_ parallelism library.
+
+        Note that the ``batch_execute`` and ``parallel`` keyword arguments are **mutually exclusive**,
+        and cannot both be set to ``True``.
+
+    **Example**
+
+    Given a simple tape transform, that replaces the :class:`~.CRX` gate with a
+    :class:`~.RY` and :class:`~.CZ` operation,
+
+    .. code-block:: python
+
+        @qml.single_tape_transform
+        def tape_transform(tape, x):
+            for op in tape.operations + tape.measurements:
+                if op.name == "CRX":
+                    qml.RY(op.parameters[0] * qml.math.sqrt(x), wires=op.wires[1])
+                    qml.CZ(wires=op.wires)
+                else:
+                    op.queue()
+
+    Let's build a QNode batch reduction that applies this transform twice with different
+    transform parameters to create two independent tapes that can be batch executed,
+    and then sums the results:
+
+    .. code-block:: python
+
+        @qml.batch_reduce
+        def my_transform(tape, x, y):
+            tape1 = tape_transform(tape, x)
+            tape2 = tape_transform(tape, y)
+
+            def processing_fn(results):
+                return qml.math.sum(results)
+
+            return [tape1, tape2], processing_fn
+
+    It can then be used to transform an existing QNode:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @my_transform(0.7, 0.8)
+        @qml.qnode(dev)
+        def circuit(x):
+            qml.Hadamard(wires=0)
+            qml.CRX(x, wires=[0, 1])
+            return qml.expval(qml.PauliZ(1))
+
+    >>> circuit(0.6)

```

Comment:
This still feels a bit weird to me. If I run `circuit`, even with the `batch_reduce` decorator, I would still instinctively expect to get the output of `circuit`, and not the processing function. Not sure if this makes sense, but is there a way to enable a distinction between the original QNode and the processing function? I guess just not using the decorator for `batch_reduce` would give the list of tapes / function explicitly?

Reply:
I think it essentially comes down to Python syntax:

- Use the decorator if you don't want to keep the original QNode, you want to replace it.
- If you want both the original QNode and the transform, don't use it as a decorator:
  
  ```python
  >>> fn = transform(qnode)
  >>> print(fn(params), qnode(params)
  ```

I think this is more of a documentation thing? E.g., `qml.grad` would work equally well as a decorator, where it would simply 'replace' the QNode:

```python
@qml.grad
def circuit(params):
```

Calling `circuit` would now return the _gradient_ of the QNode.

Reply:
So maybe in the documentation we just avoid the decorator syntax?